8275063: Implementation of Foreign Function & Memory API (Second incubator)

mcimadamore opened the pull request:
This PR contains the API and implementation changes for JEP-419 [1]. A more detailed description of such changes, to avoid repetitions during the review process, is included as a separate comment.

[1] - https://openjdk.java.net/jeps/419

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275063](https://bugs.openjdk.java.net/browse/JDK-8275063): Implementation of Foreign Function & Memory API (Second incubator)


### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to be0dd36ecd19fc2656670608146b1ad3df3f41a0
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to 7cf4fcd9d6277964ff183cae7328baa37e8c948b
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to 9fafb2a656a2db98053d4fddd0557b030cf6ee27
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5907/head:pull/5907` \
`$ git checkout pull/5907`

Update a local copy of the PR: \
`$ git checkout pull/5907` \
`$ git pull https://git.openjdk.java.net/jdk pull/5907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5907`

View PR using the GUI difftool: \
`$ git pr show -t 5907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5907.diff">https://git.openjdk.java.net/jdk/pull/5907.diff</a>

</details>

mcimadamore commented:
/csr needed

PaulSandoz commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/Utils.java:
```diff
@@ -104,34 +105,43 @@ public static long bitsToBytesOrThrow(long bits, Supplier<RuntimeException> exFa
-        if (carrier == MemoryAddress.class) {
+    public static VarHandle makeMemoryAccessVarHandle(ValueLayout layout, boolean skipAlignmentCheck) {
+        class VarHandleCache {
+            private static final Map<ValueLayout, VarHandle> handleMap = new ConcurrentHashMap<>();
+            private static final Map<ValueLayout, VarHandle> handleMapNoAlignCheck = new ConcurrentHashMap<>();
```
Something to consider later if this is an issue. Since the number of `ValueLayout` instances is fixed, carrier x order = 18, we can use stable arrays with ordinals on the instances.

mcimadamore replied:
What about alignment?

PaulSandoz replied:
Drat, `skipAlignmentCheck` misled me but perhaps there is still benefit for common constants with 8 bit and size alignment and fallback otherwise.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceScopeImpl.java:
```diff
@@ -91,72 +89,37 @@ void addInternal(ResourceList.ResourceCleanup resource) {
-
-        void release();
+    @Override
+    public void keepAlive(ResourceScope target) {
+        if (target == this) {
+            throw new IllegalArgumentException("Invalid target scope.");
+        }
+        ResourceScopeImpl targetImpl = (ResourceScopeImpl)target;
+        targetImpl.acquire0();
+        addCloseAction(targetImpl::release0);
```
Maybe this should explicitly check if target is `null` (though the call to `acquire0` would also produce an NPE, the stack trace having Objects::requireNonNull in there would make the error more obvious I think).
```suggestion
    public void keepAlive(ResourceScope target) {
        Objects.requireNonNull(target);
        if (target == this) {
            throw new IllegalArgumentException("Invalid target scope.");
        }
        ResourceScopeImpl targetImpl = (ResourceScopeImpl)target;
        targetImpl.acquire0();
        addCloseAction(targetImpl::release0);
```

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SharedScope.java:
```diff
@@ -90,18 +88,30 @@ public HandleImpl acquire() {
+    public void release0() {
+        int value;
+        do {
+            value = (int) STATE.getVolatile(jdk.internal.foreign.SharedScope.this);
```
Doesn't need to be fully qualified I think?
```suggestion
            value = (int) STATE.getVolatile(this);
```

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SharedScope.java:
```diff
@@ -90,18 +88,30 @@ public HandleImpl acquire() {
+                //cannot get here - we can't close segment twice
+                throw new IllegalStateException("Already closed");
+            }
+        } while (!STATE.compareAndSet(jdk.internal.foreign.SharedScope.this, value, value - 1));
```
Same here
```suggestion
        } while (!STATE.compareAndSet(this, value, value - 1));
```

JornVernee reviewed (approved):
Mostly some minor javadoc comments.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ArenaAllocator.java:
```diff
@@ -51,99 +60,48 @@ MemorySegment trySlice(long bytesSize, long bytesAlignment) {
+                if (maxPossibleAllocationSize > blockSize) {
                     // too big
+                    allocatedSize = Utils.alignUp(bytesSize, bytesAlignment);
+                    if (size > arenaSize) {
+                        throw new OutOfMemoryError();
+                    }
```
I'll have to think some more about this. I don't think this is covered inside the block - that is, the block tries to allocate, and then in the finally we throw if we realized we've allocated too much.

mcimadamore replied:
What is missing, I think, is a check (size > arenaSize) at the beginning of the method (we only check this in one of the paths). But we need to check before and after, I think, as it is possible to allocate a segment and then realize that we ended up overflowing the arena size.

mcimadamore replied:
While what I said above correctly reflects what the implementation does, I think a broader issue is that the arena allocator implementation is allocating sometimes more native memory than what its contract specifies. While in some cases we can prevent that, I think in the general case (e.g. where we allocate a new block) we cannot, unless we add extra API guarantees - e.g. that the arena size should be a multiple of the block size (but then we'd have to special case `Long.MAX_VALUE`, or maybe pick a "big enough" power of two instead)

JornVernee replied:
Maybe we should not support block size in the case of a bounded arena. i.e. just allocate the whole thing upfront, and have 3 APIs:

1. arena with no bounds and default block size.
2. arena with no bounds and custom block size.
3. arena with bounds, that has no blocks size but allocates the whole thing in one go (could be modeled as block size = arena size).

Right now we have 1. and 2., but instead of 3. we have a variant that allows setting both the arena size and block size.

If we want to keep what we currently have, I'd suggest changing the arena size to a block count for the variant that takes both the arena size and the block size.

Any ways, that seems like something we should address for 19 maybe, not now.

AlanBateman commented on src/java.base/share/classes/java/lang/Module.java:
```diff
@@ -109,8 +110,10 @@
-    // true, if this module allows restricted native access
-    private volatile boolean enableNativeAccess;
+    // true, if this module allows restricted native access; @Stable makes sure that modules that allow native
+    // access capture this property as a constant.
```
Do you mind fixing this comment to avoid the really long line, it sticks out compare to everything else around it.

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/IOUtil.java:
```diff
@@ -475,15 +475,15 @@ static long read(FileDescriptor fd, ByteBuffer[] bufs, int offset, int length,
     private static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
 
-    static Scope.Handle acquireScope(ByteBuffer bb, boolean async) {
+    static Runnable acquireScope(ByteBuffer bb, boolean async) {
```
At some point (not this PR) we should move the "async" out of this file, IOUtil was for synchronous I/O.

mcimadamore commented:
I have made some minor API changes (added two methods to `CLinker` to return the upcall and downcall method types, as suggested offline by @PaulSandoz). I've also cleaned up the `CLinker` javadoc, and added a section on safety consideration, streamlined the links in the package-level javadoc and renamed `MemorySegment::ofAddressNative` to simply `MemorySegment::ofAddress` (which is consistent with restricted factories in `NativeSymbol` and `VaList`).

javadoc: http://cr.openjdk.java.net/~mcimadamore/JEP-419/v2/javadoc/jdk/incubator/foreign/package-summary.html
specdiff: http://cr.openjdk.java.net/~mcimadamore/JEP-419/v2/specdiff_out/overview-summary.html

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -54,32 +55,29 @@
  * <li>if {@code L} is a {@link ValueLayout} with carrier {@code E} then there are two cases:
  *     <ul>
- *         <li>if {@code L} occurs in a parameter position and {@code E} is {@code MemoryAddress.class},
+ *         <li>if {@code L} occurs in a parameter position and {@code E} is {@code NativeAddress.class},
```
This looks spurious

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -115,6 +113,29 @@
+ *</ul>
+ * <p>
+ * Upcall stubs are generally safer to work with, as the linker runtime can validate the type of the target method
+ * handle against the provided function descriptor and report an error if any mismatch is detected. If the target method
```
But, in the case of upcalls, errors can still occur if the native code casts the pointer to the upcall stub to an incorrect type, e.g. `FunctionDescriptor.ofVoid(ADDRESS, ADDRESS)`, but on the native side cast it to `void (*)(void*)`, meaning the second argument would be garbage on the Java side. i.e. there is still room for a mismatch the same as with downcalls.

mcimadamore replied:
Yes and no. In a downcall, you just don't know what signature the downcall will feature in the native lib. So you pass a function descriptor and you hope it's ok. In the upcall case you _do_ know the signature of the Java upcall code you want to call, so you can validate the descriptor against that. Of course the native code can still cast things around in ways that blow things up, but the two problems seem somewhat different, at least to me. But I can tweak the text a bit.

JornVernee replied:
Ok, thanks.

I think of it more like this: in both cases we specify a native type as well as a Java type, both in the form of a FunctionDescriptor, from which we then derive the Java type in the form of a MethodType. If there is a mismatch here with what the native code does we are in trouble, this seems the same for downcalls and upcalls. In both cases we know the Java side for sure, it's the native side we can't validate (they are just flipped around for upcalls).

But, for upcalls there is an additional thing that can go wrong: the type of the target MethodHandle we pass could have a mismatch with the type we inferred from the FunctionDescriptor, so there we need to do an extra check. i.e. in a way this seems _less_ safe (though a different kind of safety), than downcalls, since there is an additional way to mess up with the linkage request, although we can catch that case.

mcimadamore commented:
/integrate
